File: apps/web/src/api/grocery.ts
import type { UnitInput } from './recipes'

export interface GroceryListItem { id: number; key: string; name: string; category: string; quantity: string | null; quantity_max: string | null; unit: UnitInput | null; recipe_names: string[]; raw_texts: string[]; origin: 'generated' | 'manual'; purchased_at: string | null }
export interface WeeklyGroceryList { week_start: string; week_end: string; planned_meals: number; items: GroceryListItem[]; history: GroceryListItem[] }

export async function getGroceryListWeek(weekStart: string): Promise<WeeklyGroceryList> {
  const response = await fetch(`/api/v1/grocery-lists/week?week_start=${weekStart}`)
  if (!response.ok) throw new Error('The grocery list could not be loaded.')
  return response.json()
}

export async function addManualGroceryItem(weekStart: string, rawText: string): Promise<GroceryListItem> {
  const response = await fetch(`/api/v1/grocery-lists/week/items?week_start=${weekStart}`, { method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify({ raw_text: rawText }) })
  if (!response.ok) throw new Error('The grocery item could not be added.')
  return response.json()
}

export async function setGroceryItemPurchased(id: number, purchased: boolean): Promise<GroceryListItem> {
  const response = await fetch(`/api/v1/grocery-lists/items/${id}/purchased`, { method: 'PATCH', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify({ purchased }) })
  if (!response.ok) throw new Error('The grocery item could not be updated.')
  return response.json()
}

File: apps/api/src/madplanner/schemas/grocery.py
from datetime import date
from decimal import Decimal

from datetime import datetime
from pydantic import BaseModel, Field

from madplanner.schemas.recipe import UnitInput


class GroceryListItem(BaseModel):
    id: int | None = None
    key: str
    name: str
    category: str
    quantity: Decimal | None
    quantity_max: Decimal | None
    unit: UnitInput | None
    recipe_names: list[str]
    raw_texts: list[str]
    origin: str = "generated"
    purchased_at: datetime | None = None


class ManualGroceryItemCreate(BaseModel):
    raw_text: str = Field(min_length=1, max_length=500)


class GroceryPurchasedUpdate(BaseModel):
    purchased: bool


class WeeklyGroceryListResponse(BaseModel):
    week_start: date
    week_end: date
    planned_meals: int
    items: list[GroceryListItem]
    history: list[GroceryListItem] = Field(default_factory=list)
